File: LEM_Pilas/Pila_Listas.cpp
#include <iostream>
using namespace std;

class Nodo {
public:
    int data;
    Nodo* next;

    Nodo(int d) {
        data = d;
        next = nullptr;
    }
};

class Pila_Listas {
private:
    Nodo* top;

public:
    Pila_Listas() {
        top = nullptr;
    }

    bool isEmpty() {
        return top == nullptr;
    }

    void push(int data) {
        Nodo* nuevo = new Nodo(data);
        nuevo->next = top;
        top = nuevo;
    }

    int pop() {
        if (isEmpty()) {
            cout << "Error: Underflow de Pila\n";
            return -1;
        }
        int dato = top->data;
        Nodo* temp = top;
        top = top->next;
        delete temp;
        return dato;
    }

    int peek() {
        if (isEmpty()) {
            cout << "La Pila esta vacia\n";
            return -1;
        }
        return top->data;
    }
};

int main() {
    Pila_Listas pila;
    pila.push(10);
    pila.push(20);
    pila.push(30);

    cout << "Elemento superior: " << pila.peek() << endl;
    cout << "Extrae elemento: " << pila.pop() << endl;
    cout << "Nuevo elemento superior: " << pila.peek() << endl;

    return 0;
}

File: LEM_Pilas/Pila_Listas.java
public class Pila_Listas {

    static class Nodo {
        int data;
        Nodo next;

        public Nodo(int d) {
            data = d;
            next = null;
        }
    }

    static class PilaListas {
        private Nodo top;

        public PilaListas() {
            top = null;
        }

        public boolean isEmpty() {
            return top == null;
        }

        public void push(int data) {
            Nodo nuevo = new Nodo(data);
            nuevo.next = top;
            top = nuevo;
        }

        public int pop() {
            if (isEmpty()) {
                System.out.println("Error: Underflow de Pila");
                return -1;
            }
            int dato = top.data;
            top = top.next;
            return dato;
        }

        public int peek() {
            if (isEmpty()) {
                System.out.println("La Pila esta vacia");
                return -1;
            }
            return top.data;
        }
    }

    public static void main(String[] args) {
        PilaListas pila = new PilaListas();
        pila.push(10);
        pila.push(20);
        pila.push(30);

        System.out.println("Elemento superior: " + pila.peek());
        System.out.println("Extrae elemento: " + pila.pop());
        System.out.println("Nuevo elemento superior: " + pila.peek());
    }
}